{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m/allen/programs/celltypes/workgroups/rnaseqanalysis/SEA-AD/Frozen Data\u001b[0m\n",
      "├── \u001b[01;32mSEAAD_DLPFC_RNAseq_Multiome_final-nuclei.2024-02-13.h5ad\u001b[0m\n",
      "├── \u001b[01;32mSEAAD_DLPFC_RNAseq_Singleome_final-nuclei.2024-02-13.h5ad\u001b[0m\n",
      "├── \u001b[01;32mSEAAD_MTG_ATACseq_Multiome_final-nuclei.2024-02-13.h5ad\u001b[0m\n",
      "├── \u001b[01;32mSEAAD_MTG_ATACseq_Singleome_final-nuclei.2024-02-13.h5ad\u001b[0m\n",
      "├── \u001b[01;32mSEAAD_MTG_RNAseq_Multiome_final-nuclei.2024-02-13.h5ad\u001b[0m\n",
      "├── \u001b[01;32mSEAAD_MTG_RNAseq_Singleome_final-nuclei.2024-02-13.h5ad\u001b[0m\n",
      "├── \u001b[01;34mSiletti\u001b[0m\n",
      "├── \u001b[01;32mSplit public RNAseq and ATACseq objects by chemistry.ipynb\u001b[0m\n",
      "├── \u001b[01;32mbalanced_SEAAD_MTG_RNAseq_Singleome_final-nuclei.2024-06-18.h5ad\u001b[0m\n",
      "├── \u001b[01;32mgenerate_balanced_adata.py\u001b[0m\n",
      "└── \u001b[01;34mworking\u001b[0m\n",
      "\n",
      "3 directories, 9 files\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['tree', '-L', '1', PosixPath('/allen/programs/celltypes/workgroups/rnaseqanalysis/SEA-AD/Frozen Data')], returncode=0)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "from schelp.utils.config import load_config\n",
    "from rich import print as rprint\n",
    "\n",
    "paths = load_config(dataset_key=\"init\")\n",
    "subprocess.run([\"tree\", \"-L\", \"1\", paths[\"frozen\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cells: 139000\n",
      "Number of supertype categories: 139\n",
      "Number of supertype available: 139\n",
      "Number of donors: 89\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>ratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Overall AD neuropathological Change</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>High</th>\n",
       "      <td>54779</td>\n",
       "      <td>0.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intermediate</th>\n",
       "      <td>35266</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low</th>\n",
       "      <td>18593</td>\n",
       "      <td>0.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Not AD</th>\n",
       "      <td>15439</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>14923</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     count  ratio\n",
       "Overall AD neuropathological Change              \n",
       "High                                 54779   0.39\n",
       "Intermediate                         35266   0.25\n",
       "Low                                  18593   0.13\n",
       "Not AD                               15439   0.11\n",
       "Reference                            14923   0.11"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Small balanced dataset with equal number of cells per supertype\n",
    "adata_bal = sc.read(paths[\"frozen\"] / \"balanced_SEAAD_MTG_RNAseq_Singleome_final-nuclei.2024-06-18.h5ad\", backed=\"r\")\n",
    "print(f\"Number of cells: {adata_bal.shape[0]}\")\n",
    "print(f\"Number of supertype categories: {adata_bal.obs['Supertype'].cat.categories.size}\")\n",
    "print(f\"Number of supertype available: {adata_bal.obs['Supertype'].unique().size}\")\n",
    "print(f\"Number of donors: {adata_bal.obs['Donor ID'].cat.categories.size}\")\n",
    "df = adata_bal.obs['Overall AD neuropathological Change'].value_counts().to_frame()\n",
    "df[\"ratio\"] = (df[\"count\"] / df[\"count\"].sum()).round(2)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cells: 151356\n",
      "Number of supertype categories: 139\n",
      "Number of supertype available: 139\n",
      "Number of donors: 28\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>ratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Overall AD neuropathological Change</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Intermediate</th>\n",
       "      <td>61378</td>\n",
       "      <td>0.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>High</th>\n",
       "      <td>56980</td>\n",
       "      <td>0.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low</th>\n",
       "      <td>23283</td>\n",
       "      <td>0.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Not AD</th>\n",
       "      <td>9715</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     count  ratio\n",
       "Overall AD neuropathological Change              \n",
       "Intermediate                         61378   0.41\n",
       "High                                 56980   0.38\n",
       "Low                                  23283   0.15\n",
       "Not AD                                9715   0.06"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "adata_multiome = sc.read(paths[\"frozen\"] / \"SEAAD_MTG_RNAseq_Multiome_final-nuclei.2024-02-13.h5ad\", backed=\"r\")\n",
    "print(f\"Number of cells: {adata_multiome.shape[0]}\")\n",
    "print(f\"Number of supertype categories: {adata_multiome.obs['Supertype'].cat.categories.size}\")\n",
    "print(f\"Number of supertype available: {adata_multiome.obs['Supertype'].unique().size}\")\n",
    "print(f\"Number of donors: {adata_multiome.obs['Donor ID'].cat.categories.size}\")\n",
    "df = adata_multiome.obs['Overall AD neuropathological Change'].value_counts().to_frame()\n",
    "df[\"ratio\"] = (df[\"count\"] / df[\"count\"].sum()).round(2)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cells: 1226855\n",
      "Number of supertype categories: 139\n",
      "Number of supertype available: 139\n",
      "Number of donors: 89\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>ratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Overall AD neuropathological Change</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>High</th>\n",
       "      <td>486272</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intermediate</th>\n",
       "      <td>296878</td>\n",
       "      <td>0.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low</th>\n",
       "      <td>157879</td>\n",
       "      <td>0.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Not AD</th>\n",
       "      <td>148523</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference</th>\n",
       "      <td>137303</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      count  ratio\n",
       "Overall AD neuropathological Change               \n",
       "High                                 486272   0.40\n",
       "Intermediate                         296878   0.24\n",
       "Low                                  157879   0.13\n",
       "Not AD                               148523   0.12\n",
       "Reference                            137303   0.11"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "adata_singleome = sc.read(paths[\"frozen\"] / \"SEAAD_MTG_RNAseq_Singleome_final-nuclei.2024-02-13.h5ad\", backed=\"r\")\n",
    "print(f\"Number of cells: {adata_singleome.shape[0]}\")\n",
    "print(f\"Number of supertype categories: {adata_singleome.obs['Supertype'].cat.categories.size}\")\n",
    "print(f\"Number of supertype available: {adata_singleome.obs['Supertype'].unique().size}\")\n",
    "print(f\"Number of donors: {adata_singleome.obs['Donor ID'].cat.categories.size}\")\n",
    "df = adata_singleome.obs['Overall AD neuropathological Change'].value_counts().to_frame()\n",
    "df[\"ratio\"] = (df[\"count\"] / df[\"count\"].sum()).round(2)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"font-weight: bold\">[</span>\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'sample_id'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Neurotypical reference'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Donor ID'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Organism'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Brain Region'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Sex'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Gender'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Age at Death'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Race (choice=White)'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Race (choice=Black/ African American)'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Race (choice=Asian)'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Race (choice=American Indian/ Alaska Native)'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Race (choice=Native Hawaiian or Pacific Islander)'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Race (choice=Unknown or unreported)'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Race (choice=Other)'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'specify other race'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Hispanic/Latino'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Highest level of education'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Years of education'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'PMI'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Fresh Brain Weight'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Brain pH'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Overall AD neuropathological Change'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Thal'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Braak'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'CERAD score'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Overall CAA Score'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Highest Lewy Body Disease'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Total Microinfarcts (not observed grossly)'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Total microinfarcts in screening sections'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Atherosclerosis'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Arteriolosclerosis'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'LATE'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Cognitive Status'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Last CASI Score'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Interval from last CASI in months'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Last MMSE Score'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Interval from last MMSE in months'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Last MOCA Score'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Interval from last MOCA in months'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'APOE Genotype'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Primary Study Name'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Secondary Study Name'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'NeuN positive fraction on FANS'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'RIN'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'cell_prep_type'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'facs_population_plan'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'rna_amplification'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'sample_name'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'sample_quantity_count'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'expc_cell_capture'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'method'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'pcr_cycles'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'percent_cdna_longer_than_400bp'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'rna_amplification_pass_fail'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'amplified_quantity_ng'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'load_name'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'library_prep'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'library_input_ng'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'r1_index'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'avg_size_bp'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'quantification_fmol'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'library_prep_pass_fail'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'exp_component_vendor_name'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'batch_vendor_name'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'experiment_component_failed'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'alignment'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Genome'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ar_id'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'bc'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Estimated_number_of_cells'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_number_of_reads'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_sequencing_saturation'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Mean_raw_reads_per_cell'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Q30_bases_in_barcode'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Q30_bases_in_read_2'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Q30_bases_in_UMI'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Percent_duplicates'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Q30_bases_in_sample_index_i1'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Q30_bases_in_sample_index_i2'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Reads_with_TSO'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Sequenced_read_pairs'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Valid_UMIs'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Valid_barcodes'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Reads_mapped_to_genome'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Reads_mapped_confidently_to_genome'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Reads_mapped_confidently_to_intergenic_regions'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Reads_mapped_confidently_to_intronic_regions'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Reads_mapped_confidently_to_exonic_regions'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Reads_mapped_confidently_to_transcriptome'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Reads_mapped_antisense_to_gene'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Fraction_of_transcriptomic_reads_in_cells'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Total_genes_detected'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Median_UMI_counts_per_cell'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'GEX_Median_genes_per_cell'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Multiome_Feature_linkages_detected'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Multiome_Linked_genes'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Multiome_Linked_peaks'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Confidently_mapped_read_pairs'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Fraction_of_genome_in_peaks'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Fraction_of_high_quality_fragments_in_cells'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Fraction_of_high_quality_fragments_overlapping_TSS'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Fraction_of_high_quality_fragments_overlapping_peaks'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Fraction_of_transposition_events_in_peaks_in_cells'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Mean_raw_read_pairs_per_cell'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Median_high_quality_fragments_per_cell'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Non-nuclear_read_pairs'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Number_of_peaks'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Percent_duplicates'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Q30_bases_in_barcode'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Q30_bases_in_read_1'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Q30_bases_in_read_2'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Q30_bases_in_sample_index_i1'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Sequenced_read_pairs'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_TSS_enrichment_score'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Unmapped_read_pairs'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'ATAC_Valid_barcodes'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Number of mapped reads'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Number of unmapped reads'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Number of multimapped reads'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Number of reads'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Number of UMIs'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Genes detected'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Doublet score'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Fraction mitochondrial UMIs'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Used in analysis'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Class confidence'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Class'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Subclass confidence'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Subclass'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Supertype confidence'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Supertype (non-expanded)'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Supertype'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Continuous Pseudo-progression Score'</span>,\n",
       "    <span style=\"color: #008000; text-decoration-color: #008000\">'Severely Affected Donor'</span>\n",
       "<span style=\"font-weight: bold\">]</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1m[\u001b[0m\n",
       "    \u001b[32m'sample_id'\u001b[0m,\n",
       "    \u001b[32m'Neurotypical reference'\u001b[0m,\n",
       "    \u001b[32m'Donor ID'\u001b[0m,\n",
       "    \u001b[32m'Organism'\u001b[0m,\n",
       "    \u001b[32m'Brain Region'\u001b[0m,\n",
       "    \u001b[32m'Sex'\u001b[0m,\n",
       "    \u001b[32m'Gender'\u001b[0m,\n",
       "    \u001b[32m'Age at Death'\u001b[0m,\n",
       "    \u001b[32m'Race \u001b[0m\u001b[32m(\u001b[0m\u001b[32mchoice\u001b[0m\u001b[32m=\u001b[0m\u001b[32mWhite\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m,\n",
       "    \u001b[32m'Race \u001b[0m\u001b[32m(\u001b[0m\u001b[32mchoice\u001b[0m\u001b[32m=\u001b[0m\u001b[32mBlack\u001b[0m\u001b[32m/ African American\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m,\n",
       "    \u001b[32m'Race \u001b[0m\u001b[32m(\u001b[0m\u001b[32mchoice\u001b[0m\u001b[32m=\u001b[0m\u001b[32mAsian\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m,\n",
       "    \u001b[32m'Race \u001b[0m\u001b[32m(\u001b[0m\u001b[32mchoice\u001b[0m\u001b[32m=\u001b[0m\u001b[32mAmerican\u001b[0m\u001b[32m Indian/ Alaska Native\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m,\n",
       "    \u001b[32m'Race \u001b[0m\u001b[32m(\u001b[0m\u001b[32mchoice\u001b[0m\u001b[32m=\u001b[0m\u001b[32mNative\u001b[0m\u001b[32m Hawaiian or Pacific Islander\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m,\n",
       "    \u001b[32m'Race \u001b[0m\u001b[32m(\u001b[0m\u001b[32mchoice\u001b[0m\u001b[32m=\u001b[0m\u001b[32mUnknown\u001b[0m\u001b[32m or unreported\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m,\n",
       "    \u001b[32m'Race \u001b[0m\u001b[32m(\u001b[0m\u001b[32mchoice\u001b[0m\u001b[32m=\u001b[0m\u001b[32mOther\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m,\n",
       "    \u001b[32m'specify other race'\u001b[0m,\n",
       "    \u001b[32m'Hispanic/Latino'\u001b[0m,\n",
       "    \u001b[32m'Highest level of education'\u001b[0m,\n",
       "    \u001b[32m'Years of education'\u001b[0m,\n",
       "    \u001b[32m'PMI'\u001b[0m,\n",
       "    \u001b[32m'Fresh Brain Weight'\u001b[0m,\n",
       "    \u001b[32m'Brain pH'\u001b[0m,\n",
       "    \u001b[32m'Overall AD neuropathological Change'\u001b[0m,\n",
       "    \u001b[32m'Thal'\u001b[0m,\n",
       "    \u001b[32m'Braak'\u001b[0m,\n",
       "    \u001b[32m'CERAD score'\u001b[0m,\n",
       "    \u001b[32m'Overall CAA Score'\u001b[0m,\n",
       "    \u001b[32m'Highest Lewy Body Disease'\u001b[0m,\n",
       "    \u001b[32m'Total Microinfarcts \u001b[0m\u001b[32m(\u001b[0m\u001b[32mnot observed grossly\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m,\n",
       "    \u001b[32m'Total microinfarcts in screening sections'\u001b[0m,\n",
       "    \u001b[32m'Atherosclerosis'\u001b[0m,\n",
       "    \u001b[32m'Arteriolosclerosis'\u001b[0m,\n",
       "    \u001b[32m'LATE'\u001b[0m,\n",
       "    \u001b[32m'Cognitive Status'\u001b[0m,\n",
       "    \u001b[32m'Last CASI Score'\u001b[0m,\n",
       "    \u001b[32m'Interval from last CASI in months'\u001b[0m,\n",
       "    \u001b[32m'Last MMSE Score'\u001b[0m,\n",
       "    \u001b[32m'Interval from last MMSE in months'\u001b[0m,\n",
       "    \u001b[32m'Last MOCA Score'\u001b[0m,\n",
       "    \u001b[32m'Interval from last MOCA in months'\u001b[0m,\n",
       "    \u001b[32m'APOE Genotype'\u001b[0m,\n",
       "    \u001b[32m'Primary Study Name'\u001b[0m,\n",
       "    \u001b[32m'Secondary Study Name'\u001b[0m,\n",
       "    \u001b[32m'NeuN positive fraction on FANS'\u001b[0m,\n",
       "    \u001b[32m'RIN'\u001b[0m,\n",
       "    \u001b[32m'cell_prep_type'\u001b[0m,\n",
       "    \u001b[32m'facs_population_plan'\u001b[0m,\n",
       "    \u001b[32m'rna_amplification'\u001b[0m,\n",
       "    \u001b[32m'sample_name'\u001b[0m,\n",
       "    \u001b[32m'sample_quantity_count'\u001b[0m,\n",
       "    \u001b[32m'expc_cell_capture'\u001b[0m,\n",
       "    \u001b[32m'method'\u001b[0m,\n",
       "    \u001b[32m'pcr_cycles'\u001b[0m,\n",
       "    \u001b[32m'percent_cdna_longer_than_400bp'\u001b[0m,\n",
       "    \u001b[32m'rna_amplification_pass_fail'\u001b[0m,\n",
       "    \u001b[32m'amplified_quantity_ng'\u001b[0m,\n",
       "    \u001b[32m'load_name'\u001b[0m,\n",
       "    \u001b[32m'library_prep'\u001b[0m,\n",
       "    \u001b[32m'library_input_ng'\u001b[0m,\n",
       "    \u001b[32m'r1_index'\u001b[0m,\n",
       "    \u001b[32m'avg_size_bp'\u001b[0m,\n",
       "    \u001b[32m'quantification_fmol'\u001b[0m,\n",
       "    \u001b[32m'library_prep_pass_fail'\u001b[0m,\n",
       "    \u001b[32m'exp_component_vendor_name'\u001b[0m,\n",
       "    \u001b[32m'batch_vendor_name'\u001b[0m,\n",
       "    \u001b[32m'experiment_component_failed'\u001b[0m,\n",
       "    \u001b[32m'alignment'\u001b[0m,\n",
       "    \u001b[32m'Genome'\u001b[0m,\n",
       "    \u001b[32m'ar_id'\u001b[0m,\n",
       "    \u001b[32m'bc'\u001b[0m,\n",
       "    \u001b[32m'GEX_Estimated_number_of_cells'\u001b[0m,\n",
       "    \u001b[32m'GEX_number_of_reads'\u001b[0m,\n",
       "    \u001b[32m'GEX_sequencing_saturation'\u001b[0m,\n",
       "    \u001b[32m'GEX_Mean_raw_reads_per_cell'\u001b[0m,\n",
       "    \u001b[32m'GEX_Q30_bases_in_barcode'\u001b[0m,\n",
       "    \u001b[32m'GEX_Q30_bases_in_read_2'\u001b[0m,\n",
       "    \u001b[32m'GEX_Q30_bases_in_UMI'\u001b[0m,\n",
       "    \u001b[32m'GEX_Percent_duplicates'\u001b[0m,\n",
       "    \u001b[32m'GEX_Q30_bases_in_sample_index_i1'\u001b[0m,\n",
       "    \u001b[32m'GEX_Q30_bases_in_sample_index_i2'\u001b[0m,\n",
       "    \u001b[32m'GEX_Reads_with_TSO'\u001b[0m,\n",
       "    \u001b[32m'GEX_Sequenced_read_pairs'\u001b[0m,\n",
       "    \u001b[32m'GEX_Valid_UMIs'\u001b[0m,\n",
       "    \u001b[32m'GEX_Valid_barcodes'\u001b[0m,\n",
       "    \u001b[32m'GEX_Reads_mapped_to_genome'\u001b[0m,\n",
       "    \u001b[32m'GEX_Reads_mapped_confidently_to_genome'\u001b[0m,\n",
       "    \u001b[32m'GEX_Reads_mapped_confidently_to_intergenic_regions'\u001b[0m,\n",
       "    \u001b[32m'GEX_Reads_mapped_confidently_to_intronic_regions'\u001b[0m,\n",
       "    \u001b[32m'GEX_Reads_mapped_confidently_to_exonic_regions'\u001b[0m,\n",
       "    \u001b[32m'GEX_Reads_mapped_confidently_to_transcriptome'\u001b[0m,\n",
       "    \u001b[32m'GEX_Reads_mapped_antisense_to_gene'\u001b[0m,\n",
       "    \u001b[32m'GEX_Fraction_of_transcriptomic_reads_in_cells'\u001b[0m,\n",
       "    \u001b[32m'GEX_Total_genes_detected'\u001b[0m,\n",
       "    \u001b[32m'GEX_Median_UMI_counts_per_cell'\u001b[0m,\n",
       "    \u001b[32m'GEX_Median_genes_per_cell'\u001b[0m,\n",
       "    \u001b[32m'Multiome_Feature_linkages_detected'\u001b[0m,\n",
       "    \u001b[32m'Multiome_Linked_genes'\u001b[0m,\n",
       "    \u001b[32m'Multiome_Linked_peaks'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Confidently_mapped_read_pairs'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Fraction_of_genome_in_peaks'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Fraction_of_high_quality_fragments_in_cells'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Fraction_of_high_quality_fragments_overlapping_TSS'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Fraction_of_high_quality_fragments_overlapping_peaks'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Fraction_of_transposition_events_in_peaks_in_cells'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Mean_raw_read_pairs_per_cell'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Median_high_quality_fragments_per_cell'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Non-nuclear_read_pairs'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Number_of_peaks'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Percent_duplicates'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Q30_bases_in_barcode'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Q30_bases_in_read_1'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Q30_bases_in_read_2'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Q30_bases_in_sample_index_i1'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Sequenced_read_pairs'\u001b[0m,\n",
       "    \u001b[32m'ATAC_TSS_enrichment_score'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Unmapped_read_pairs'\u001b[0m,\n",
       "    \u001b[32m'ATAC_Valid_barcodes'\u001b[0m,\n",
       "    \u001b[32m'Number of mapped reads'\u001b[0m,\n",
       "    \u001b[32m'Number of unmapped reads'\u001b[0m,\n",
       "    \u001b[32m'Number of multimapped reads'\u001b[0m,\n",
       "    \u001b[32m'Number of reads'\u001b[0m,\n",
       "    \u001b[32m'Number of UMIs'\u001b[0m,\n",
       "    \u001b[32m'Genes detected'\u001b[0m,\n",
       "    \u001b[32m'Doublet score'\u001b[0m,\n",
       "    \u001b[32m'Fraction mitochondrial UMIs'\u001b[0m,\n",
       "    \u001b[32m'Used in analysis'\u001b[0m,\n",
       "    \u001b[32m'Class confidence'\u001b[0m,\n",
       "    \u001b[32m'Class'\u001b[0m,\n",
       "    \u001b[32m'Subclass confidence'\u001b[0m,\n",
       "    \u001b[32m'Subclass'\u001b[0m,\n",
       "    \u001b[32m'Supertype confidence'\u001b[0m,\n",
       "    \u001b[32m'Supertype \u001b[0m\u001b[32m(\u001b[0m\u001b[32mnon-expanded\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m,\n",
       "    \u001b[32m'Supertype'\u001b[0m,\n",
       "    \u001b[32m'Continuous Pseudo-progression Score'\u001b[0m,\n",
       "    \u001b[32m'Severely Affected Donor'\u001b[0m\n",
       "\u001b[1m]\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rprint(adata_bal.obs.columns.tolist())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scgpt",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
